41-vaccine-temp-monitor/firmware/arduino/src/main.cpp

```diff
@@ -0,0 +1,528 @@
+ *                          |           |
+ *                          | V G S S   |
+ *                          | C N D C O |
+ *                          |_C_D_A_L_S_|
+ * __________________,        | | | | |        ,__________________
+ *       ESP32       |        | | | | |        |   NOTECARRIER-B
+ *                   |        | | | | |        |
+ *                   |        | | | | '--------|AUX1
+ *                   |        | | | |          |
+ *                   |        '-^-^-^----------|<VIO
+ *                   |          | | |          |
+ *               VBAT|----------^-^-^----------|VBAT
+ *                   |          | | |          |
+ *                GND|----------'-^-^----------|GND
+ *                   |            | |          |
+ *                SDA|------------'-^----------|SDA
+ *                   |              |          |
+ *                SCL|--------------'----------|SCL
+ *                   |                         |
+ *                 EN|-------------------------|ATTN
+ *                   |                         |
+ */
+
+#include <M2M_LM75A.h>
+#include <Notecard.h>
+#include <NotecardEnvVarManager.h>
+#ifdef ARDUINO_ARCH_ESP32
+#include <NotecardAuxiliaryWiFi.h>
+#endif // ARDUINO_ARCH_ESP32
+
+// Uncomment this line and replace com.your-company:your-product-name with your
+// ProductUID.
+// #define PRODUCT_UID "com.your-company:your-product-name"
+
+#ifndef PRODUCT_UID
+#define PRODUCT_UID "" // "com.my-company.my-name:my-project"
+#pragma message "PRODUCT_UID is not defined in this example. Please ensure your Notecard has a product identifier set before running this example or define it in code here. More details at https://bit.ly/product-uid"
+#endif
+#define usbSerial Serial
+#define txRxPinsSerial Serial1
+
+#define CRT_MAX_TEMP_C 30
+#define CRT_SAFE_TEMP_C 25
+#define DEFAULT_SAMPLE_INTERVAL_S 1800
+#define RETRY_S 5
+
+M2M_LM75A lm75a;
+Notecard notecard;
+NotecardEnvVarManager *envVarManager = NULL;
+#ifdef ARDUINO_ARCH_ESP32
+using namespace blues;
+NotecardAuxiliaryWiFi aux_wifi(notecard);
+#endif
+
+volatile int alarm_hysteresis = CRT_SAFE_TEMP_C;
+volatile int alarm_temperature = CRT_MAX_TEMP_C;
+volatile size_t sample_interval_s = DEFAULT_SAMPLE_INTERVAL_S;
+
+void logNoteF(const char *format_, ...)
+{
+    char log[256];
+    va_list args;
+    va_start(args, format_);
+    vsnprintf(log, sizeof(log), format_, args);
+
+    if (J *cmd = notecard.newCommand("hub.log"))
+    {
+        JAddStringToObject(cmd, "text", log);
+        if (!notecard.sendRequest(cmd))
+        {
+            JDelete(cmd);
+        }
+    }
+
+    va_end(args);
+}
+
+int acquireGPSLocation(size_t timeout_s_ = 95)
+{
+    int result;
+    const char *gps_mode;
+    size_t gps_time_s;
+
+    // Get Current Configuration
+    if (J *rsp = notecard.requestAndResponse(notecard.newRequest("card.location.mode")))
+    {
+        gps_mode = JGetString(rsp, "mode");
+        NoteDeleteResponse(rsp);

```

Comment:
Doesn't deleting the response invalidate the gps_mode pointer? I think you would need to copy the string out before deleting.

Reply:
Great catch! Absolutely right.

Not sure what I was thinking. If it was making a copy, then I would have been leaking. :shrug: